### PR TITLE
[FEAT] Add a `list.contains` API

### PR DESCRIPTION
**Pull request by raunakab**
# Overview
Added a `list.contains` API to check whether or not a column of type `List` or `FixedSizeList` contains a given value.

### CODE REVIEW

File: src/daft-functions/src/list/contains.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        "list_contains"
+    }
+
+    fn to_field(&self, inputs: &[ExprRef], schema: &Schema) -> DaftResult<Field> {
+        match inputs {
+            [data, value] => {
+                let data_field = data.to_field(schema)?;
+                let value_field = value.to_field(schema)?;
+                match &data_field.dtype {
+                    DataType::List(inner_type) => {
```
Comment (kevinzwang) [inline src/daft-functions/src/list/contains.rs:30]: Another check you should do with `inner_type` is to check if it is a type that you can check equality with the values.

**Review by kevinzwang — CHANGES_REQUESTED** [review]
An alternative way to write the kernel is to check equality between the flattened list and the values and then for each list range, determine if there are any true values in that equality array. The potential advantage of this is that you can use the [vectorized arrow2 equality op](https://docs.rs/arrow2/latest/arrow2/compute/comparison/fn.eq.html). The potential disadvantage is that you can't early terminate if you find an equal value in a list. 

Also, please add tests!

File: src/daft-functions/src/list/contains.rs (hunk 1 above)
Comment (kevinzwang) [reply src/daft-functions/src/list/contains.rs:30]: Also, the dtype can be a `FixedSizeList`

**Comment by samster25** [comment]
@raunakab @kevinzwang status on this PR?

File: daft/expressions/expressions.py [hunk 2]
```diff
@@ -2911,6 +2911,17 @@ def join(self, delimiter: str | Expression) -> Expression:
         delimiter_expr = Expression._to_expression(delimiter)
         return Expression._from_pyexpr(self._expr.list_join(delimiter_expr._expr))
 
+    def contains(self, value: Expression) -> Expression:
+        """Returns all lists which contain the given value.
+
+        Args:
+            value (DataType): the value to search for in the list
```
[the remark this answers is not in the document]
Comment (raunakab) [reply daft/expressions/expressions.py:3017]: Updated!

File: src/daft-core/src/array/ops/list.rs [hunk 3]
```diff
@@ -475,6 +476,31 @@ impl ListArray {
             self.validity().cloned(),
         ))
     }
+
+    pub fn list_contains(&self, values: Series) -> DaftResult<BooleanArray> {
+        assert_eq!(self.len(), values.len(), "Expected two lists with the same length, instead got self.len() = {} and values.len() = {}", self.len(), values.len());
```
[the remark this answers is not in the document]
Comment (raunakab) [reply src/daft-core/src/array/ops/list.rs:631]: Updated.

File: daft/expressions/expressions.py [hunk 4]
```diff
@@ -2911,6 +2911,17 @@ def join(self, delimiter: str | Expression) -> Expression:
...
+
+        Returns:
+            Expression: a List expression which are the lists which all contain at least one instance of the given value.
+        """
+        return Expression._from_pyexpr(native.list_contains(self._expr, value._expr))
```
[the remark this answers is not in the document]
Comment (raunakab) [reply daft/expressions/expressions.py:3022]: How about `self` then?

[the remark this answers is not in the document]
Comment (kevinzwang) [reply daft/expressions/expressions.py:3022]: Self is already an expression so there's no need. It should look something like
```py
value_expr = Expression.to_expression(value)
return Expression._from_pyexpr(native.list_contains(self._expr, value_expr._expr))
```

[the remark this answers is not in the document]
Comment (raunakab) [reply daft/expressions/expressions.py:3022]: Wait, I don't understand.

`value` is already of type `Expression`. `self` is of type `ExpressionListNamespace`. When I grab `self._expr` and `value._expr`, I am grabbing their internal `PyExpr` type, no? That is what the `native.list_contains` function expects.

[the remark this answers is not in the document]
Comment (raunakab) [reply daft/expressions/expressions.py:3022]: Why should we perform the `Expression._to_expression` conversion first? Looking at the implementation, it seems to take in a `object` and return the input if it is of type `Expression`.

File: daft/expressions/expressions.py [hunk 5]
```diff
@@ -2911,6 +2911,17 @@ def join(self, delimiter: str | Expression) -> Expression:
...
         return Expression._from_pyexpr(self._expr.list_join(delimiter_expr._expr))
 
+    def contains(self, value: Expression) -> Expression:
+        """Returns all lists which contain the given value.
+
+        Args:
+            value (DataType): the value to search for in the list
+
+        Returns:
+            Expression: a List expression which are the lists which all contain at least one instance of the given value.
```
[the remark this answers is not in the document]
Comment (raunakab) [reply daft/expressions/expressions.py:3020]: I believe the original issue that @Vince7778 requested was to have the output be a column of lists in which each list contains the value that was being searched for in the first place.

The original issue gives a solid example, I believe.

[the remark this answers is not in the document]
Comment (raunakab) [reply daft/expressions/expressions.py:3020]: I actually looked through my implementation. I implemented it incorrectly. Will update it.

[the remark this answers is not in the document]
Comment (jaychia) [reply daft/expressions/expressions.py:3020]: Please take a closer look at the original issue.

1. The returned Expression is a boolean expression, indicating whether the list in each row contains the provided value
2. In the example provided, this boolean expression is then provided to a `df.where` which is performing the filtering.

There should be no concept of filtering at all in this PR. This is a simple scalar expression.

**Comment by raunakab** [comment]
This is a stale PR. Closing in favour of https://github.com/Eventual-Inc/Daft/pull/3479.

**Pull request closed without merging**